leveldb-server/src/test/db.match.spec.ts

```diff
@@ -0,0 +1,175 @@
+  const caseInsensitive = pattern.flags.includes('i');
+  const filter = { operator: 'matches' as 'matches', path: ['value'], pattern: pattern.source, caseInsensitive };
+  t.true(match({ key: 'abc', value: testValue }, filter));
+};
+stringMatches.title = (providedTitle = '', testValue, pattern) =>
+  `${providedTitle} ${JSON.stringify(testValue)} =~ ${pattern})`;
+
+const stringDoesntMatch: Macro<[any, RegExp]> = (t, testValue, pattern) => {
+  const caseInsensitive = pattern.flags.includes('i');
+  // TypeScript type system not strong enough to deduce this always
+  // creates a legal filter, so just cast.
+  const filter = { operator: 'matches' as 'matches', path: ['value'], pattern: pattern.source, caseInsensitive };
+  t.false(match({ key: 'abc', value: testValue }, filter));
+};
+stringDoesntMatch.title = (providedTitle = '', testValue, matchValue) =>
+  `${providedTitle} ${JSON.stringify(testValue)} !~ ${matchValue})`;
+
+test('matches matches string', stringDoesntMatch, null, /abc/i);
+test('matches matches string', stringDoesntMatch, 0, /abc/i);
+test('matches matches string', stringMatches, 'abc', /abc/i);
+test('matches matches string', stringMatches, 'ABC', /abc/i);
+test('matches matches string', stringDoesntMatch, 'ABC', /abc/);
+
+const startsWith: Macro<[any, string]> = (t, testValue, value) => {
+  const filter = { operator: 'startsWith' as 'startsWith', path: ['value'], value };
+  t.true(match({ key: 'abc', value: testValue }, filter));
+};
+startsWith.title = (providedTitle = '', testValue, value) =>
+  `${providedTitle} ${testValue}.startsWith(${value})`;
+
+const doesntStartWith: Macro<[any, string]> = (t, testValue, value) => {
+  const filter = { operator: 'startsWith' as 'startsWith', path: ['value'], value };
+  t.false(match({ key: 'abc', value: testValue }, filter));
+};
+doesntStartWith.title = (providedTitle = '', testValue, value) =>
+  `${providedTitle} !${testValue}.startsWith(${value})`;
+
+test('startsWith matches string', doesntStartWith, null, 'ab');
+test('startsWith matches string', doesntStartWith, 0, 'ab');
+test('startsWith matches string', startsWith, 'abc', 'ab');
+test('startsWith matches string', doesntStartWith, 'ABC', 'ab');
+
+const startsWithAb = { operator: 'startsWith' as 'startsWith', path: ['key'], value: 'ab' };
+const equalsAb = { operator: 'eq' as 'eq', path: ['key'], value: 'ab' };
+const equals0 = { operator: 'eq' as 'eq', path: ['value'], value: 0 };
+const equals1 = { operator: 'eq' as 'eq', path: ['value'], value: 1 };
+
+test('and applies all filters', (t) => {
+  t.true(match({ key: 'abc', value: 0 }, { operator: 'and', filters: [startsWithAb, equals0] }));
+  t.false(match({ key: 'abc', value: 0 }, { operator: 'and', filters: [startsWithAb, equals1] }));
+  t.false(match({ key: 'abc', value: 0 }, { operator: 'and', filters: [equalsAb, equals0] }));
+  t.false(match({ key: 'abc', value: 0 }, { operator: 'and', filters: [equalsAb, equals1] }));
```

Comment:
consider adding match with filter of same path
`t.true(match({ key: 'ab', value: 0 }, { operator: 'and', filters: [startsWithAb, equalsAb] }));`
and
`t.true(match({ key: 'abc', value: 0 }, { operator: 'and', filters: [startsWithAb, equalsAb] }));`

Reply:
I find @bergundy 's version (mostly copied above) to be easier to read, because the 2 parts are quite different.